src/engine/database/mod.rs

```diff
@@ -134,6 +135,86 @@ pub trait DatabaseEndpoint {
     async fn pool(&self) -> Result<DatabasePool, Error>;
 }
 
+/// Represents the different type of comparison match operations
+#[derive(Clone, Debug)]
+pub enum Operation {
+    EQ,
+    CONTAINS,
+    IN,
+    GT,
+    GTE,
+    LT,
+    LTE
+}
+
+/// Struct representing a value comparison. In query operations, visitors take provided 
+/// operation/value nested map and converted them into a `Comparison` struct and pass
+/// it on the database-specific transaction for use in creating match portion of queries. 
+#[derive(Clone, Debug)]
+pub struct Comparison {
+    operation: Operation,
+    operand: Value,
+    negated: bool
+}
+
+impl Comparison {
+
+    pub fn new(operation: Operation, negated: bool, operand: Value) -> Self {
+        Comparison {
+            operation,
+            operand,
+            negated
+        }
+    }
+
+    pub fn default(v: Value) -> Self {
+        Self::new(Operation::EQ, false, v)
+    }
+
+}
+
+impl TryFrom<Value> for Comparison {
+
+    type Error = Error;
+
+    fn try_from(v: Value) -> Result<Comparison, Error> {
+        Ok(match v {
+            Value::String(_) => Comparison::default(v),
+            Value::Int64(_) => Comparison::default(v),
+            Value::Float64(_) => Comparison::default(v),
+            Value::Bool(_) => Comparison::default(v),
+            Value::Map(mut m) => {
+                let operation_str = m.keys().next()
```

Comment:
I think what you want here is something more like

```
let (operation_str, operand) = m.into_iter().next()
```

The consuming form of the iterator, `into_iter` as opposed to `iter` (or just `keys`, as in the code now), returns owned values rather than references, avoiding the need to clone. That does consume the source, but I think that should be fine in this context, as type conversations do consume source values.

Reply:
I knew you would know the answer :)